### PR TITLE
[TEST] Performance test for Vienna format

### LINKED ISSUE #1392: [Structure IO] Format Vienna Benchmark
**Level 1 - Easy**

- [x] reading from stream
- [x] reading from file

See format_fasta_benchmark.cpp as a reference.

### PR DESCRIPTION
**Pull request by joergi-w**
Adds performance tests for the Vienna structure format.

Note: The run time for SeqAn2 is significantly slower, because the structure representation is a graph instead of a string over the wuss alphabet.

Resolves #1392.

### CODE REVIEW

File: test/performance/io/format_vienna_benchmark.cpp [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+    #include <seqan/rna_io.h>
+#endif
+
+#include <sstream>
+#include <string>
+
+#include <benchmark/benchmark.h>
+
+#include <seqan3/io/structure_file/input.hpp>
+#include <seqan3/io/structure_file/input_format_concept.hpp>
```
Comment (marehr) [inline test/performance/io/format_vienna_benchmark.cpp]: do you need these headers?

File: test/performance/io/format_vienna_benchmark.cpp [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+#include <seqan3/io/structure_file/output.hpp>
+#include <seqan3/io/structure_file/output_format_concept.hpp>
```
Comment (marehr) [inline test/performance/io/format_vienna_benchmark.cpp]: do you need these headers?

Comment (joergi-w) [reply test/performance/io/format_vienna_benchmark.cpp]: I need the input/output header, but not the concept header. I remove the latter.

File: test/performance/io/format_vienna_benchmark.cpp [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+{
+    "(((((((..((((........)))).((((.........)))).....(((((.......))))))))))))......."
+    "(((((((..((((........)))).((((.........)))).....(((((.......))))))))))))......."
+    "(((((((..((((........)))).((((.........)))).....(((((.......))))))))))))......."
+    "(((((((..((((........)))).((((.........)))).....(((((.......))))))))))))......."
+    "(((((((..((((........)))).((((.........)))).....(((((.......))))))))))))......."
+    "(((((((..((((........)))).((((.........)))).....(((((.......))))))))))))......."
+};
+
+static std::string vienna_file = []()
```
Comment (h-2) [inline test/performance/io/format_vienna_benchmark.cpp]:
```suggestion
inline std::string const vienna_file = []()
```

?

File: test/performance/io/format_vienna_benchmark.cpp [hunk 4]
```diff
@@ -0,0 +1,149 @@
+// -----------------------------------------------------------------------------------------------------
+// Copyright (c) 2006-2019, Knut Reinert & Freie Universität Berlin
+// Copyright (c) 2016-2019, Knut Reinert & MPI für molekulare Genetik
+// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
+// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE.md
+// -----------------------------------------------------------------------------------------------------
+
+#if __has_include(<seqan/rna_io.h>)
```
Comment (smehringer) [inline test/performance/io/format_vienna_benchmark.cpp]:
```suggestion
#if SEQAN3_HAS_SEQAN2
```
we have a macro for this :)

File: test/performance/io/format_vienna_benchmark.cpp [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+#include <seqan3/io/structure_file/format_vienna.hpp>
+#include <seqan3/test/performance/units.hpp>
+
+using namespace seqan3;
+using namespace seqan3::test;
+
+inline constexpr size_t iterations_per_run = 1024;
+
+inline std::string const header{"seq foobar blobber"};
+inline std::string const sequence
```
Comment (smehringer) [inline test/performance/io/format_vienna_benchmark.cpp]: We have the utility `generate_sequence` in `test/include/seqan3/test/performance/sequence_generator`

```cpp
#include <seqan3/test/performance/sequence_generator.hpp>
...
auto dna4_seq seq = generate_sequence<dna4>(500 /*length*/, 0/*variance of uniform dist*/, 0 /*seed*/)
                  | seqan3::views::to_char 
                  | seqan3::to<std::string>;
```
:)

File: test/performance/io/format_vienna_benchmark.cpp [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+
+inline std::string const vienna_file = []()
+{
+    std::string file{};
+    for (size_t idx = 0; idx < iterations_per_run; idx++)
+        file += "> " + header + "\n" + sequence + "\n" + structure + "\n";
+    return file;
+}();
+
+void write3(benchmark::State & state)
```
Comment (smehringer) [inline test/performance/io/format_vienna_benchmark.cpp]: Can we call it `write_seqan3`? write3 is not very expressive (I know you probably copied it from another benchmark)

Comment (joergi-w) [reply test/performance/io/format_vienna_benchmark.cpp]: True, good idea!

File: test/performance/io/format_vienna_benchmark.cpp (hunk 5 above)
Comment (joergi-w) [reply test/performance/io/format_vienna_benchmark.cpp]: with views::persist this works well now...

**Comment by smehringer** [comment]
let's wait for CI

**Pull request merged**